### ISSUE TITLE
arc4random, why it isn't inplemented?

### ISSUE THREAD

**Issue by patricus3** [jart/cosmopolitan#1343]
Hi.
I have two questions, why you didn't inplement arc4random?
and an other, the discord drama all was a big misunderstanding, check the new issue at superconfigure repo, and see what I actually meant.

**Comment by jart** [comment]
Cosmopolitan provides a variety of ways to generate random numbers, such as getrandom() and getentropy() which are both recommended for all use cases. No one's found need of arc4random yet, which is why we don't have it. I'm perfectly happy to include that too, if someone wants to copy over the FreeBSD or OpenBSD implementation per the CONTRIBUTING.md process.

**Comment by patricus3** [comment]
@jart ok, , cool
but what about the second problem? I want to discuss cosmopolitan on discord and you banned me because of the stupid misunderstanding. we'ree supposed to cooperate as a community, and I am willing to submit small patches here and there soon if I will be welcome back to the community.

**Comment by mrdomino** [comment]
Some resources:
* [{Free,Open}BSD arc4random.c][0] (the libc interface for both is much the same)
* However the [FreeBSD kernel arc4random][1] includes [Fenestras X][2], which is based on
* [Niels Ferguson’s Windows 10 RNG whitepaper][3]

[0]: https://github.com/openbsd/src/blob/8c0ef5529f7d57aeee5bacadf4c844a997adf7f7/lib/libc/crypt/arc4random.c
[1]: https://github.com/freebsd/freebsd-src/blob/d830cac9fed1882a192b0ec5c96fb3ac9bfbc2ee/sys/libkern/arc4random.c
[2]: https://www.freebsd.org/status/report-2020-10-2020-12.html#Fenestras-X-random(4)
[3]: https://download.microsoft.com/download/1/c/9/1c9813b8-089c-4fef-b2ad-ad80e79403ba/Whitepaper%20-%20The%20Windows%2010%20random%20number%20generation%20infrastructure.pdf

**Comment by mrdomino** [comment]
TLDR on Fenestras X + the Windows 10 RNG is:

* The kernel maintains one root randomness pool per logical core so that no synchronization is needed
* Random numbers are generated via AES-CTR, which directly services large requests or else fills up a 128-byte buffer that is pulled from for small requests (to amortize the cost of the AES key expansion.)
* The libc arc4random calls use `MAP_INHERIT_ZERO` (spelled `INHERIT_ZERO` in FreeBSD) to reset the buffer state on forks.

There are lots more subtleties in the whitepaper, and also some interesting design choices in Fenestras X in FreeBSD.

_However_, I think the `arc4random` family of syscalls might actually be implemented on all the platforms we care about (in Windows it's just spelled `ProcessPrng`) so at least the initial implementation of this might be pretty boring.

**Comment by mrdomino** [comment]
Okay, right, they're not syscalls, they're in libc. So there's at least some porting work to be done.
